src/S3/Input/CopyObjectRequest.php

```diff
@@ -859,4 +861,17 @@ public function requestUri(): string
 
         return "/{$uri['Bucket']}/{$uri['Key']}";
     }
+
+    public function validate(): void
+    {
+        foreach (['Bucket', 'CopySource', 'Key'] as $name) {
+            if (null === $value = $this->$name) {
+                throw new InvalidArgument(sprintf('Missing parameter "%s" when validating the "%s". The value cannot be null.', $name, __CLASS__));
+            }
+
+            if (\is_object($value) && method_exists($value, 'validate')) {
```

Comment:
that could be guessed at build time isn't it?

Reply:
Could.. but it is a bit tricky.. I rather leave that for a future improvement.